### DISCUSSION TITLE
Is there a way to access files in the repo without the ssh and/or git protocol?

### DISCUSSION THREAD

**Discussion by QuantumLibet** [charmbracelet/soft-serve#455] [Q&A] [answered]
Allow me to start by stating that I have a feeling that I'm missing the obvious. You know, when you have that uneasy feeling that your question is rather dumb, since the solution is so obvious, yet you don't see it. So you have to ask. That's me. Now.

Please let me know if you can provide any insight. I have not opened a discussion before and may be misunderstanding its purpose. Please close this discussion, if inappropriate.

<br>

Currently, I store the public keys of SSH users in a Gitea repository and fetch them during login using the `AuthorizedKeysCommand` in the `sshd_config`.

This method works well with `curl` and the HTTPS protocol.

<br>

I would like to migrate this repo to `soft`, like all the others. However, I am unable to access the files in the repository without an authorized SSH user or the `git` executable in combination with a repo token. There is neither a fully authorized SSH user, nor the `git` binary installed at this point in time.

<br>

Despite my attempts to use more basic tools such as curl, wget and rsync (including the syntax of: rsync -e "ssh -p 23231 -l repo_user_name" ...), I have been unsuccessful.

Since `soft` supports HTTP[S] and allows for anonymous and keyless access, there may be a simple solution that I am not seeing.

<br>
<br>

the gist:
example: it is assumed that `soft` is installed and configured, with an entry `soft-serve` in the current users' `~/.ssh/config`. 

```
mkdir test_repo && cd test_repo
echo "fake public key content" > fake_public_key.pub
git init && git add * && git commit -m 'first entry'
git remote add origin soft-serve:test_repo
git push --set-upstream origin main
ssh soft-serve settings anon-access true
ssh soft-serve settings allow-keyless true
```

How can I download fake_public_key.pub from the repo without git or ssh?

I would appreciate any ideas or feedback, even if it is just to confirm that what I am striving for is not possible.
Thank you for your time.

**Comment by aymanbagabas** [comment, marked as answer]
Hi @QuantumLibet, thank you for opening this discussion.

Currently, only the SSH server supports the `blob` API interface. Adding a new HTTP endpoint for those APIs makes sense when dealing with a single file and a `text/plain` header. I've created a new issue https://github.com/charmbracelet/soft-serve/issues/456 to track this feature.

For now, you can use the `blob` SSH command as an unauthenticated user using `keyboard-interactive` SSH auth type `ssh -o PreferredAuthentications=keyboard-interactive <host> -- repo blob <repo> <optional git revision> <file path>`

**Reply by QuantumLibet** [reply]
Thank you very much for your quick response and the most helpful way possible, by even creating a new feature request!
I am impressed, even more so, as I assumed that my enquiry would be rejected, as it has nothing to do with the actual purpose of soft as a git repo. I am all the more pleased (outright happy!) about your answer and the possibility of now being able to move the last repo!

This year couldn't start any better. A heartfelt Thank You!

tl;dr
AWESOME! 👍🏻
